### PR TITLE
Revert optimizer_step() nan error changes

### PR DESCRIPTION
**Pull request by glenn-jocher**
@Laughing-q

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Simplifies the training optimizer step by removing legacy PyTorch 1.9 handling and unifying gradient clipping and scaler behavior ⚙️

### 📊 Key Changes
- Removed `TORCH_1_9` import and version-specific branches in `trainer.py`.
- Always apply gradient clipping with `torch.nn.utils.clip_grad_norm_(..., max_norm=10.0)`.
- Removed try/except around optimizer stepping; now directly calls `self.scaler.step(self.optimizer)` and `self.scaler.update()`.
- Relies on PyTorch GradScaler’s built-in handling for non-finite gradients instead of custom warnings.

### 🎯 Purpose & Impact
- Streamlines and modernizes the training loop, focusing on PyTorch 2.x (notably 2.4+) ✅
- Reduces complexity and maintenance overhead by removing legacy code 🧹
- Keeps training stability intact; GradScaler still skips optimizer steps when encountering NaNs/Infs, but without the previous custom warning ⚠️
- Produces cleaner logs and may offer minor performance benefits due to fewer conditional checks 🚀
- Note: Implicitly de-emphasizes compatibility with very old PyTorch versions (e.g., 1.9).

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline reviews. An Ultralytics engineer will also assist you here shortly 🙂

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional PR-specific notes:
- 🔁 Behavior Confirmation: Since this PR adjusts `optimizer_step()` handling of non-finite gradients and AMP, please confirm expected behavior for both AMP enabled/disabled, and that training proceeds consistently across supported PyTorch versions.
- 🧪 Tests: Consider adding a unit/integration test that exercises non-finite gradient scenarios (e.g., extreme hyperparameters) to validate the new flow with `GradScaler.step(optimizer)` and `clip_grad_norm_`.
- 📝 Changelog/Docs: If this materially changes user-facing behavior (e.g., warnings removed, step-skipping now fully delegated to AMP), please note it in the [documentation](https://docs.ultralytics.com/).
- 🧩 Backward Compatibility: Removing version-specific logic (e.g., `TORCH_1_9`) simplifies code—please verify no regressions for supported environments in CI.

If this PR addresses a bug (e.g., NaN errors), please include a Minimum Reproducible Example (MRE) in your description:
1. Exact environment details (Python, PyTorch/CUDA, OS, GPU)
2. Minimal script or CLI command that reproduces the issue on `main` and demonstrates behavior with this PR
3. Clear expected vs. actual results

Example MRE template (adjust as needed):
```python
from ultralytics import YOLO

# Replace with a small, public dataset config like 'coco128.yaml' if applicable
model = YOLO('yolo11n.pt')  # or your model
model.train(
    data='coco128.yaml',
    epochs=1,
    imgsz=640,
    amp=True,  # also test amp=False
    lr0=0.01,
    device=0,
)
```

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by UltralyticsAssistant** [comment]
Absolutely thrilled to see this merged—thank you, @glenn-jocher! As Antoine de Saint-Exupéry said, “Perfection is achieved not when there is nothing more to add, but when there is nothing left to take away.” This PR beautifully embodies that spirit by simplifying the optimizer step, unifying gradient clipping and GradScaler behavior, and modernizing our training loop for PyTorch 2.x.

The result: cleaner logs, less maintenance overhead, and a sharper path forward for YOLO11 and beyond. Appreciated and impactful work!

See details in PR [Streamline optimizer step and remove legacy branches](https://github.com/ultralytics/ultralytics/pull/22360).
